sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -52,6 +52,12 @@ object CollationTypeCasts extends TypeCoercionRule {
       overlay.withNewChildren(collateToSingleType(Seq(overlay.input, overlay.replace))
         ++ Seq(overlay.pos, overlay.len))
 
+    case stringPadExpr@(_: StringLPad | _: StringRPad) =>
+      val newChildren = collateToSingleType(
+        Seq(stringPadExpr.children.head, stringPadExpr.children(2)))
+      stringPadExpr.withNewChildren(
+        Seq(newChildren.head, stringPadExpr.children(1), newChildren(1)))
```

Comment:
```suggestion
      val Seq(str, pad) = collateToSingleType(Seq(stringLPad.str, stringLPad.pad))
      stringLPad.withNewChildren(Seq(str, stringLPad.len, pad))
```

Spark expressions have parameter names, so we might as well use them

Reply:
```suggestion
      val Seq(str, len, pad) = stringPadExpr.children
      val Seq(newStr, newPad) = collateToSingleType(Seq(str, pad))
      stringPadExpr.withNewChildren(Seq(newStr, len, newPad))
```

note: the previous approach will require having 2 separate cases
```
case stringLPad: StringLPad =>
case stringRPad: StringRPad =>
```

but if you do want to preserve the single case, try something like the second approach